<!DOCTYPE html>
<html lang="en">

  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
 
    <title>Hooke</title>
    <meta name="description" content="Hooke - A powerful software toolkit for differential expression analysis.">
 
<link rel="shortcut icon" type="image/png" href="/hooke/images/favicon.png"/> 
   <link rel="stylesheet" href="/hooke/css/main.css">
   <link rel="stylesheet" href="/hooke/node_modules/prismjs/themes/base16-atelierlakeside.light.css" />
   <link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.css">
 
   <link rel="canonical" href="http://localhost:4000/hooke/docs/projection/">
   
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-QW21JE7HZ7"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-QW21JE7HZ7');
</script>

     <!-- Begin of scripts to run the widget -->
     <script src="/hooke/widget_scripts/htmlwidgets-1.2/htmlwidgets.js"></script>
     <link href="/hooke/widget_scripts/rglwidgetClass-2/rgl.css" rel="stylesheet" />
     <script src="/hooke/widget_scripts/rglwidgetClass-2/rglClass.src.js"></script>
     <script src="/hooke/widget_scripts/CanvasMatrix4-2016/CanvasMatrix.src.js"></script>
     <script src="/hooke/widget_scripts/rglWebGL-binding-0.99.16/rglWebGL.js"></script>
     <script src="/hooke/widget_scripts/jquery-1.11.3/jquery.min.js"></script>
     <link href="/hooke/widget_scripts/crosstalk-1.0.0/css/crosstalk.css" rel="stylesheet" /> 
     <script src="/hooke/widget_scripts/crosstalk-1.0.0/js/crosstalk.min.js"></script> 
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/4.1.1/js/bootstrap.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
     <!-- End of scripts to run the widget -->
 </head>






  <body data-spy= "scroll" data-target= "#toc">
    <div class="container-fullwidth">
    <nav class="navbar navbar-inverse navbar-static-top">
      <div class="container">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="/hooke/">Hooke</a>
        </div>
        <div id="navbar" class="collapse navbar-collapse">
          <ul class="nav navbar-nav">
              <li><a href="/hooke/docs/introduction">Documentation</a></li>
              <!-- <li><a href="/hooke/Paul/">Paul</a></li> -->
              <li><a href="/hooke/contributors/">Contributors</a></li>
              <li><a href="/hooke/papers/">Publications</a></li>
              <li><a href="https://github.com/cole-trapnell-lab/hooke">GitHub</a></li>     
          </ul>


        <script async src="https://cse.google.com/cse.js?cx=002844359906969258106:dsowufscyiq"></script>
        <div class="gcse-searchresults-only"></div>
        <link rel="stylesheet" href="/hooke/css/search.css">
        <form action="" method="GET">
            <input class="form-control mr-sm-2" class="input" name="q" type="search" placeholder="Search site" aria-label="Search">
        </form>
      </div>
      </div>
    </nav>
    </div>
  </header>
  

    
    <div class= "container">
  <div class= "row">
    <div class= "col-sm-4">
      <ul class="section-nav" id= "toc" data-spy= "affix" data-toggle= "toc">
        <li class="toc-entry toc-h2"><a href="/hooke/docs/introduction">Introduction</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/updates">Major updates in Monocle 3</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/installation">Installing Monocle 3</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/help">Getting help</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/starting">Getting started with Monocle 3</a></li> 
        <li class="toc-entry toc-h2"><a href="/hooke/docs/alignment">Pre-process your data</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/clustering">Clustering and classifying your cells</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/trajectories">Constructing single-cell trajectories</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/differential">Differential expression analysis</a></li>
        <li class="toc-entry toc-h2"><a class="active" href="/hooke/docs/projection">Projecting query data onto a reference</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/store_monocle_objects">Storing monocle objects</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/cicero">Single-cell ATAC-seq data</a></li>
        <li class="toc-entry toc-h2"><a href="/hooke/docs/citations">Citations and acknowledgments</a></li>
      </ul>
    </div>
    <!-- main content area start -->
    <div class= "col-sm-8">
      <nav aria-label="breadcrumb">
        <ol class="breadcrumb">
          <li class="breadcrumb-item"><a href="/hooke/">Home</a></li>
          <li class="breadcrumb-item"><a href="/hooke/docs/introduction">Docs</a></li>
          <li class="breadcrumb-item active" aria-current="page">Project query data onto a reference.</li>
        </ol>
      </nav>

<h2>Project a query data set onto a reference data set</h2>

<p>
Co-embedding is used to compare similar data sets in order to identify similarities and differences and transfer annotations between cells. When the data sets are large and there are many sets to compare, the memory and run time requirements for processing co-embedded data can become impediments. Monocle3's solution is to save the models that transform a reference data set to low-dimensional PCA and UMAP space, and then use the models to transform query data sets to the low-dimensional reference space where one can plot them for comparison or annotate query cells by finding similar reference cells.
</p>

<h3 id="load-datasets">Load the reference and query data sets</h3>

<p>
We begin by loading the reference and query data sets into Monocle3. To simplify this example, we load both data sets together; however, one can process them independently.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">library</span><span class="p">(</span><span class="n">monocle3</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span><span class="w">

</span><span class="c1"># Load the reference data set.
</span><span class="n">matrix_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readMM</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/cao.mouse_embryo.sample.mtx.gz"</span><span class="p">)))</span><span class="w">
</span><span class="n">cell_ann_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/cao.mouse_embryo.sample.coldata.txt.gz"</span><span class="p">),</span><span class="w"> </span><span class="n">text</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">
</span><span class="n">gene_ann_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/cao.mouse_embryo.sample.rowdata.txt.gz"</span><span class="p">),</span><span class="w"> </span><span class="n">text</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">

</span><span class="n">cds_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">new_cell_data_set</span><span class="p">(</span><span class="n">matrix_ref</span><span class="p">,</span><span class="w">
                             </span><span class="n">cell_metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cell_ann_ref</span><span class="p">,</span><span class="w">
                             </span><span class="n">gene_metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_ann_ref</span><span class="p">)</span><span class="w">

</span><span class="c1"># Load the query data set.
</span><span class="n">matrix_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readMM</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/srivatsan.mouse_embryo_scispace.sample.mtx.gz"</span><span class="p">)))</span><span class="w">
</span><span class="n">cell_ann_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/srivatsan.mouse_embryo_scispace.sample.coldata.txt.gz"</span><span class="p">),</span><span class="w"> </span><span class="n">text</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">
</span><span class="n">gene_ann_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.csv</span><span class="p">(</span><span class="n">gzcon</span><span class="p">(</span><span class="n">url</span><span class="p">(</span><span class="s2">"https://depts.washington.edu:/trapnell-lab/software/monocle3/mouse/data/srivatsan.mouse_embryo_scispace.sample.rowdata.txt.gz"</span><span class="p">),</span><span class="w"> </span><span class="n">text</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s1">'\t'</span><span class="p">)</span><span class="w">

</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">new_cell_data_set</span><span class="p">(</span><span class="n">matrix_qry</span><span class="p">,</span><span class="w">
                             </span><span class="n">cell_metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cell_ann_qry</span><span class="p">,</span><span class="w">
                             </span><span class="n">gene_metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene_ann_qry</span><span class="p">)</span></code></pre></figure>


<h3 id="remove-non-shared-genes">Remove genes that are not in both data sets</h3>

<p>
It is essential that the query cds has the same genes in the same order as the reference cds, so we identify the shared genes and sort them.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Genes in reference.
</span><span class="n">genes_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">row.names</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">)</span><span class="w">

</span><span class="c1"># Genes in query.
</span><span class="n">genes_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">row.names</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)</span><span class="w">

</span><span class="c1"># Shared genes.
</span><span class="n">genes_shared</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">intersect</span><span class="p">(</span><span class="n">genes_ref</span><span class="p">,</span><span class="w"> </span><span class="n">genes_qry</span><span class="p">)</span><span class="w">

</span><span class="c1"># Remove non-shared genes.
</span><span class="n">cds_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cds_ref</span><span class="p">[</span><span class="n">genes_shared</span><span class="p">,]</span><span class="w">
</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cds_qry</span><span class="p">[</span><span class="n">genes_shared</span><span class="p">,]</span></code></pre></figure>


<h3 id="umi-cutoff">Apply a common UMI cutoff</h3>

<p>
We filter the cells by applying the same UMI cutoff to our data sets. For these data sets, we applied a cutoff of 1000 to reduce their sizes but we show how to find these cutoffs.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Reference data set UMI cutoff.
</span><span class="n">numi_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">colData</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">)[[</span><span class="s1">'Total_mRNAs'</span><span class="p">]])</span><span class="w">
</span><span class="c1"># numi_ref is 1001
</span><span class="n">numi_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">min</span><span class="p">(</span><span class="n">colData</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)[[</span><span class="s1">'n.umi'</span><span class="p">]])</span><span class="w">
</span><span class="err">#</span><span class="w"> </span><span class="n">numi_qry</span><span class="w"> </span><span class="n">is</span><span class="w"> </span><span class="m">1000</span></code></pre></figure>


<h3 id="estimate-size-factors">Estimate size factors</h3>

<p>
After applying the gene and UMI filters, we re-calculate the size factors of both data sets.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">cds_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">estimate_size_factors</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">)</span><span class="w">
</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">estimate_size_factors</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)</span></code></pre></figure>


<h3 id="process-reference">Process the reference data set</h3>

<p>
We are ready to process the reference data set by transforming it into PCA and and UMAP low dimension spaces. By setting the <code>build_nn_index</code> to TRUE, we build a nearest neighbor index in the UMAP space, which we will use to transfer the reference annotations to the query. After processing, we save the transform models and nearest neighbor index so that we can use them to transform the query data set into the reference space.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">cds_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">preprocess_cds</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">,</span><span class="w"> </span><span class="n">num_dim</span><span class="o">=</span><span class="m">100</span><span class="p">)</span><span class="w">
</span><span class="n">cds_ref</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduce_dimension</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">,</span><span class="w"> </span><span class="n">build_nn_index</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="c1"># Save the PCA and UMAP transform models for use with projection.
</span><span class="n">save_transform_models</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">,</span><span class="w"> </span><span class="s1">'cds_ref_test_models'</span><span class="p">)</span></code></pre></figure>


<h3 id="project-query">Project the query data set into the reference space</h3>

<p>
Our query data set was loaded and filtered already, so we can use the <code>preprocess_transform()</code> and <code>reduce_dimension_transform()</code> functions to transform the query into the reference space. The <code>load_transform_models()</code> function loads the reference transform models into the query cds where they are used by <code>preprocess_transform()</code> and <code>reduce_dimension_transform()</code>.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Load the reference transform models into the query cds.
</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">load_transform_models</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">,</span><span class="w"> </span><span class="s1">'cds_ref_test_models'</span><span class="p">)</span><span class="w">
</span><span class="c1"># Apply the reference transform models to the query cds.
</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">preprocess_transform</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)</span><span class="w">
</span><span class="n">cds_qry</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reduce_dimension_transform</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)</span></code></pre></figure>


<h3 id="project-query">Plot the combined data sets</h3>

<p>
First we plot the reference and query cells in UMAP space.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">plot_cells</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">)</span><span class="w">
</span><span class="n">plot_cells</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)</span></code></pre></figure>


      <div class= "text-center">
        <img src= "/hooke/images/manual_images/L2_projection_reference.png" width= 450>
      </div>

      <div class= "text-center">
        <img src= "/hooke/images/manual_images/L2_projection_query.png" width= 450>
      </div>


<h3 id="project-query">Plot the combined data sets</h3>

<p>
Now we label the cells in the reference and query cdses, combine the cdses, and plot the combined cells.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="c1"># Label the data sets.
</span><span class="n">colData</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">)[[</span><span class="s1">'data_set'</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'reference'</span><span class="w">
</span><span class="n">colData</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">)[[</span><span class="s1">'data_set'</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s1">'query'</span><span class="w">
</span><span class="c1"># Combine the reference and query data sets.
</span><span class="n">cds_combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">combine_cds</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">,</span><span class="w"> </span><span class="n">cds_qry</span><span class="p">),</span><span class="w">  </span><span class="n">keep_all_genes</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">cell_names_unique</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">keep_reduced_dims</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">plot_cells</span><span class="p">(</span><span class="n">cds_combined</span><span class="p">,</span><span class="w"> </span><span class="n">color_cells_by</span><span class="o">=</span><span class="s1">'data_set'</span><span class="p">)</span></code></pre></figure>


      <div class= "text-center">
        <img src= "/hooke/images/manual_images/L2_projection_combined.png" width= 450>
      </div>


<h3 id="transfer-cell-labels-query">Transfer the reference cell labels to the query data set</h3>

<p>
Now we transfer the cell type annotations from the reference to the query data set using the nearest neighbor index that we made in the reference UMAP space.
</p>

<figure class="highlight"><pre><code class="language-r" data-lang="r"><span class="n">cds_qry_lab_xfr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transfer_cell_labels</span><span class="p">(</span><span class="n">cds_qry</span><span class="p">,</span><span class="w"> </span><span class="n">reduction_method</span><span class="o">=</span><span class="s1">'UMAP'</span><span class="p">,</span><span class="w"> </span><span class="n">ref_coldata</span><span class="o">=</span><span class="n">colData</span><span class="p">(</span><span class="n">cds_ref</span><span class="p">),</span><span class="w"> </span><span class="n">ref_column_name</span><span class="o">=</span><span class="s1">'Main_cell_type'</span><span class="p">,</span><span class="w"> </span><span class="n">query_column_name</span><span class="o">=</span><span class="s1">'cell_type_xfr'</span><span class="p">,</span><span class="w"> </span><span class="n">transform_models_dir</span><span class="o">=</span><span class="s1">'cds_ref_test_models'</span><span class="p">)</span><span class="w">
</span><span class="n">cds_qry_lab_fix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fix_missing_cell_labels</span><span class="p">(</span><span class="n">cds_qry_lab_xfr</span><span class="p">,</span><span class="w"> </span><span class="n">reduction_method</span><span class="o">=</span><span class="s1">'UMAP'</span><span class="p">,</span><span class="w"> </span><span class="n">from_column_name</span><span class="o">=</span><span class="s1">'cell_type_xfr'</span><span class="p">,</span><span class="w"> </span><span class="n">to_column_name</span><span class="o">=</span><span class="s1">'cell_type_fix'</span><span class="p">)</span></code></pre></figure>

<!-- main content area end -->
</div></div></div>


    <script src="/hooke/node_modules/jquery/dist/jquery.min.js"></script>
<script src="/hooke/node_modules/bootstrap-sass/assets/javascripts/bootstrap.min.js"></script>
<script src="/hooke/node_modules/prismjs/prism.js"></script>
<script src="/hooke/node_modules/prismjs/prism.r.js" type="text/javascript"></script>
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v0.4.1/dist/bootstrap-toc.min.js"></script>

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
</script>
<script type="text/javascript" async
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML">
</script>

<footer>

  <div class="wrapper">
      <p>&copy; 2022 Cole Trapnell</p>
  </div>

</footer>







  </body>

</html>