---
description: "Papers describing Monocle."
modified: 2013-09-11
tags: [Monocle]
---

<div class="container">
  <div class="row">
    <!-- sidebar, which will move to the top on a small screen -->
    <div class="col-sm-4">
      <nav id="toc" data-spy="affix" data-toggle="toc"></nav>
    </div>
    <!-- main content area -->
    <div class="col-sm-12">


    <div class="text-center">
      <h3 data-toc-skip>Monocle is an ongoing research project as well as a toolkit. If you use Monocle, <strong>please cite these papers</strong> in your work! </h3>
    </div>
<br>

<hr class="featurette-divider">

<div class="row featurette">
  <div class="col-md-3">
    <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/trapnell-cacchiarelli-monocle.png" alt="Monocle 1">
  </div>

  <div class="col-md-9">
    <h3 data-toc-skip>The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells</h3>
    <p>Cole Trapnell*, Davide Cacchiarelli*, Jonna Grimsby, Prapti Pokharel, Shuqiang Li, Michael Morse, Niall J. Lennon, Kenneth J. Livak, Tarjei S. Mikkelsen, John L. Rinn.
    <p><em>Nature Biotechnology</em> 2014
    <br>
    <p> Defining the transcriptional dynamics of a temporal process such as cell differentiation is challenging owing to the high variability in gene expression between individual cells. Time-series gene expression analyses of bulk cells have difficulty distinguishing early and late phases of a transcriptional cascade or identifying rare subpopulations of cells, and single-cell proteomic methods rely on a priori knowledge of key distinguishing markers. Here we describe Monocle, an unsupervised algorithm that increases the temporal resolution of transcriptome dynamics using single-cell RNA-Seq data collected at multiple time points. Applied to the differentiation of primary human myoblasts, Monocle revealed switch-like changes in expression of key regulatory factors, sequential waves of gene regulation, and expression of regulators that were not known to act in differentiation. We validated some of these predicted regulators in a loss-of function screen. Monocle can in principle be used to recover single-cell gene expression kinetics from a wide array of cellular processes, including differentiation, proliferation and oncogenic transformation
    <p><strong>This is the original Monocle paper, which introduced the concept of pseudotime ordering for single-cell analysis</strong>

    <div class="text-center">
      <a class="btn btn-default btn-sm" href="http://dx.doi.org/10.1038/nbt.2859" role="button">HTML</a>
      <a class="btn btn-default btn-sm" href="http://cole-trapnell-lab.github.io/pdfs/papers/trapnell-cacchiarelli-monocle.pdf" role="button">PDF</a>
    </div>

  </div>
</div>

<hr class="featurette-divider">

<div class="row featurette">
  <div class="col-md-3">
    <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/census-BEAM2.png" alt="Census + BEAM">
  </div>

  <div class="col-md-9">
    <h3 data-toc-skip>Single-cell mRNA quantification and differential analysis with Census</h3>
    <p>Xiaojie Qiu, Andrew Hill,	Jonathan Packer, Dejun Lin,	Yi-An Ma, Cole Trapnell
    <p><em>Nature Methods</em> 2017
    <br>
    <p> Single-cell gene expression studies promise to reveal rare cell types and cryptic states, but the high variability of single-cell RNA-seq measurements frustrates efforts to assay transcriptional differences between cells. We introduce the Census algorithm to convert relative RNA-seq expression levels into relative transcript counts without the need for experimental spike-in controls. Analyzing changes in relative transcript counts led to dramatic improvements in accuracy compared to normalized read counts and enabled new statistical tests for identifying developmentally regulated genes. Census counts can be analyzed with widely used regression techniques to reveal changes in cell-fate-dependent gene expression, splicing patterns and allelic imbalances. We reanalyzed single-cell data from several developmental and disease studies, and demonstrate that Census enabled robust analysis at multiple layers of gene regulation. Census is freely available through our updated single-cell analysis toolkit, Monocle 2.

    <p><strong>This paper describes BEAM (used in branch analysis) and Census (the core of <code>relative2abs</code>)</strong>

    <div class="text-center">
      <a class="btn btn-default btn-sm" href="http://dx.doi.org/10.1038/nmeth.4150" role="button">HTML</a>
      <a class="btn btn-default btn-sm" href="http://cole-trapnell-lab.github.io/pdfs/papers/qiu-census.pdf" role="button">PDF</a>
    </div>

  </div>
</div>

<hr class="featurette-divider">

<div class="row featurette">
  <div class="col-md-3">
    <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/monocle2.png" alt="Monocle 2">
  </div>

  <div class="col-md-9">
    <h3 data-toc-skip>Reversed graph embedding resolves complex single-cell trajectories</h3>
    <p>Xiaojie Qiu, Qi Mao, Ying Tang, Li Wang,	Raghav Chawla, Hannah Pliner, Cole Trapnell
    <p><em>Nature Methods</em> 2017

    <br>
    <p> Single-cell trajectories can unveil how gene regulation governs cell fate decisions. However, learning the structure of complex trajectories with multiple branches remains a challenging computational problem. We present Monocle 2, an algorithm that uses reversed graph embedding to describe multiple fate decisions in a fully unsupervised manner. We applied Monocle 2 to two studies of blood development and found that mutations in the genes encoding key lineage transcription factors divert cells to alternative fates.

    <p><strong>This paper describes Monocle 2 and the use of Reversed Graph Embedding for single-cell analysis.</strong>

    <div class="text-center">
      <a class="btn btn-default btn-sm" href="https://www.nature.com/articles/nmeth.4402" role="button">HTML</a>
      <a class="btn btn-default btn-sm" href="http://cole-trapnell-lab.github.io/pdfs/papers/qiu-monocle2.pdf" role="button">PDF</a>
    </div>

  </div>
</div>

<hr class="featurette-divider">

<div class="row featurette">
  <div class="col-md-3">
    <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/cao-plot.png" alt="Monocle 2">
  </div>

  <div class="col-md-9">
    <h3 data-toc-skip>The single-cell transcriptional landscape of mammalian organogenesis</h3>
    <p>Junyue Cao, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell, and Jay Shendure
    <p><em>Nature 2019</em>

    <br>
    <p> Mammalian organogenesis is a remarkable process. Within a short timeframe, the cells of the three germ layers transform into an embryo that includes most of the major internal and external organs. Here we investigate the transcriptional dynamics of mouse organogenesis at single-cell resolution. Using single-cell combinatorial indexing, we profiled the transcriptomes of around 2 million cells derived from 61 embryos staged between 9.5 and 13.5 days of gestation, in a single experiment. The resulting ‘mouse organogenesis cell atlas’ (MOCA) provides a global view of developmental processes during this critical window. We use Monocle 3 to identify hundreds of cell types and 56 trajectories, many of which are detected only because of the depth of cellular coverage, and collectively define thousands of corresponding marker genes. We explore the dynamics of gene expression within cell types and trajectories over time, including focused analyses of the apical ectodermal ridge, limb mesenchyme and skeletal muscle.

    <div class="text-center">
      <a class="btn btn-default btn-sm" href="https://www.nature.com/articles/s41586-019-0969-x?WT.ec_id=NATURE-201902&sap-outbound-id=3D6B78CB95C498488E0AB2A140F1B7A311607DA1" role="button">HTML</a>
      <a class="btn btn-default btn-sm" href="https://cole-trapnell-lab.github.io/pdfs/papers/cao-spielmann-mouse-emb.pdf" role="button">PDF</a>
    </div>

  </div>
</div>

</div>
</div>
</div>
