
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Maddy Duran">
      
      
        <link rel="canonical" href="https://cole-trapnell-lab.github.io/hooke/advanced/">
      
      
        <link rel="prev" href="../about/">
      
      
        <link rel="next" href="../reference/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Time Series Analysis - Hooke</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="trapnell" data-md-color-primary="indigo" data-md-color-accent="teal">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#analyzing-time-series-perturbation-data-with-hooke" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Hooke" class="md-header__button md-logo" aria-label="Hooke" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Hooke
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Time Series Analysis
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/cole-trapnell-lab/hooke" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Hooke" class="md-nav__button md-logo" aria-label="Hooke" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Hooke
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/cole-trapnell-lab/hooke" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Background
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_type/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Type Requirements
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Example Datasets
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Differential Abundance
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Time Series Analysis
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Time Series Analysis
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#analysis-of-cranial-sensory-ganglia-with-hooke" class="md-nav__link">
    <span class="md-ellipsis">
      Analysis of cranial sensory ganglia with Hooke
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analysis of cranial sensory ganglia with Hooke">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fitting-a-hooke-model-with-model-time-points" class="md-nav__link">
    <span class="md-ellipsis">
      Fitting a Hooke model with model time points
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-hooke-time-intervals-to-view-kinetics" class="md-nav__link">
    <span class="md-ellipsis">
      Using Hooke time intervals to view kinetics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using Hooke time intervals to view kinetics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#controlling-for-batch-effects" class="md-nav__link">
    <span class="md-ellipsis">
      Controlling for batch effects
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perturbation-kinetics" class="md-nav__link">
    <span class="md-ellipsis">
      Perturbation kinetics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#analysis-of-cranial-sensory-ganglia-with-hooke" class="md-nav__link">
    <span class="md-ellipsis">
      Analysis of cranial sensory ganglia with Hooke
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analysis of cranial sensory ganglia with Hooke">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fitting-a-hooke-model-with-model-time-points" class="md-nav__link">
    <span class="md-ellipsis">
      Fitting a Hooke model with model time points
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-hooke-time-intervals-to-view-kinetics" class="md-nav__link">
    <span class="md-ellipsis">
      Using Hooke time intervals to view kinetics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using Hooke time intervals to view kinetics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#controlling-for-batch-effects" class="md-nav__link">
    <span class="md-ellipsis">
      Controlling for batch effects
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perturbation-kinetics" class="md-nav__link">
    <span class="md-ellipsis">
      Perturbation kinetics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="analyzing-time-series-perturbation-data-with-hooke">Analyzing time series perturbation data with Hooke</h1>
<p>Hooke can be used to analyze data with multiple time points across perturbations, such as the data found in 
<a href="https://www.biorxiv.org/content/10.1101/2022.08.04.502764v1">Saunders, Srivatsan, et al. Nature, in press (2023)</a>. This study in includes ~3 million single cells across almost 2000 individual barcoded embryos. It includes 19 timepoints (18-96 hpf) and 23 genetic loss of function experiments. For more information see the <a href="https://cole-trapnell-lab.github.io/zscape/">ZSCAPE website</a>. </p>
<h3 id="analysis-of-cranial-sensory-ganglia-with-hooke">Analysis of cranial sensory ganglia with Hooke</h3>
<p>This vignette focuses on the cranial sensory ganglia subset. Cranial sensory neurons transmit information from the head, ear, heart and viscera. Around 30,000 cranial sensory neurons (~20 cells/embryo) were identified. These cells were annotated by comparing branch specific gene expression with published expression data.</p>
<p><img alt="csg_umap" src="../assets/csg_umap.png" width="75%" /></p>
<h4 id="fitting-a-hooke-model-with-model-time-points">Fitting a Hooke model with model time points</h4>
<p><em>Foxi1</em> and <em>phox2a</em> are two transcription factors important for cranial sensory ganglia development. Here we subset the <code>cds</code> to focus on those perturbations. We also include all the controls in our <code>cell count set</code> construction. </p>
<pre><code>
control_ids = c(&quot;ctrl-uninj&quot;, &quot;ctrl-inj&quot;, &quot;ctrl-hgfa&quot;, &quot;ctrl-met&quot;, &quot;ctrl-mafba&quot;, &quot;ctrl-noto&quot;, &quot;ctrl-tbx16&quot;)
cds = cds[, colData(cds)$gene_target %in% c(&quot;phox2a&quot;, &quot;foxi1&quot;, control_ids)]

colData(cds)$perturbation = ifelse(colData(cds)$gene_target %in% control_ids, 
                                   &quot;control&quot;, 
                                   colData(cds)$gene_target) 

ccs = new_cell_count_set(cds, 
                         sample_group = &quot;embryo&quot;, 
                         cell_group = &quot;cell_type_sub&quot;)
</code></pre>
<p>The function<code>build_interval_formula()</code> builds a model formula for time series models based on the range of the data. This is a utility function that puts the knots in reasonable positions based on the range of the data. It takes the following as input: </p>
<ul>
<li><code>ccs</code>- a Hooke <code>cell_count_set</code> object. </li>
<li><code>num_breaks</code> - specifies degrees of freedom</li>
<li><code>interval_var</code> - name of the column that the interval should be based on </li>
<li><code>interval_start</code>- defaults to the <code>min(interval_var)</code> value</li>
<li><code>interval_stop</code> - defaults to the <code>max(interval_var)</code> value</li>
</ul>
<p><em>Note: if using this function or specifying your own spline, you must load <code>library(spline)</code>. Putting <code>splines::ns()</code> in your model formula will result in an error as the model reads it as an interaction term.</em></p>
<pre><code>start_time = 18
stop_time = 72
time_formula = build_interval_formula(ccs, num_breaks = 3, interval_start = 18, interval_stop = 72)
</code></pre>
<p>The resulting time formula is <code>~ ns(timepoint , knots= c(45))</code></p>
<p>We can fit a <code>cell_count_model</code> using this time formula and a perturbation term... </p>
<pre><code>ccm = new_cell_count_model(ccs, 
                           main_model_formula_str = paste0(time_formula, &quot;+ perturbation&quot;))

# predict for 48 hpf 
cond_wt = estimate_abundances(ccm, tibble(timepoint = 48, perturbation = &quot;control&quot;))
cond_phox2a = estimate_abundances(ccm, tibble(timepoint = 48, perturbation = &quot;phox2a&quot;))
cond_foxi1 = estimate_abundances(ccm, tibble(timepoint = 48, perturbation = &quot;foxi1&quot;))

wt_v_phox2a_tbl = compare_abundances(ccm, cond_wt, cond_phox2a)
wt_v_foxi1_tbl = compare_abundances(ccm, cond_wt, cond_foxi1)
</code></pre>
<p>... and plot the abundance changes. </p>
<pre><code>plot_contrast(ccm, wt_v_phox2a_tbl, x=1, y=3, q_value_thresh = 0.05)
</code></pre>
<p><img alt="phox2a_48hpf" src="../assets/phox2a_48hpf.png" width="75%" /></p>
<h4 id="using-hooke-time-intervals-to-view-kinetics">Using Hooke time intervals to view kinetics</h4>
<p>We can use Hooke to view when cell types are at their peak abundance in the dataset. Here we are just looking at wild-type data and estimating abundances over a time interval. </p>
<pre><code>
wt_cds = cds[, colData(cds)$gene_target %in% control_ids]
wt_ccs = new_cell_count_set(wt_cds, 
                           sample_group = &quot;embryo&quot;, 
                           cell_group = &quot;cell_type_sub&quot;)

colData(wt_ccs)$timepoint = as.numeric(colData(wt_ccs)$timepoint)
wt_ccm = new_cell_count_model(wt_ccs, 
                              main_model_formula_str = &quot;ns(timepoint, df=3)&quot;)

wt_timepoint_pred_df = estimate_abundances_over_interval(wt_ccm, 
                                                         interval_start=18, 
                                                         interval_stop=72, 
                                                         interval_col=&quot;timepoint&quot;, 
                                                         interval_step=2)

log_abund_detection_thresh=-3

ggplot(wt_timepoint_pred_df, aes(x = timepoint)) +
  geom_line(aes(y = exp(log_abund) + exp(log_abund_detection_thresh))) +
  facet_wrap(~cell_group, scales=&quot;free_y&quot;) + monocle3:::monocle_theme_opts()

</code></pre>
<p><img alt="wt_kinetics" src="../assets/wt_kinetics.png" /></p>
<h5 id="controlling-for-batch-effects">Controlling for batch effects</h5>
<p>If the data was collected in multiple batches, you can also include a <code>nuisance_model_formula_str</code> and plot each experimental batch's kinetics. </p>
<pre><code>
wt_expt_ccm = new_cell_count_model(wt_ccs, 
                                   main_model_formula_str = &quot;ns(timepoint, df=3)&quot;, 
                                   nuisance_model_formula_str = &quot;~ expt&quot;)

batches = data.frame(batch = unique(colData(wt_ccs)$expt))                                   
batches = batches %&gt;% mutate(tp_preds = purrr::map(.f = function(batch) {
 estimate_abundances_over_interval(wt_expt_ccm,
                                            start_time,
                                            stop_time,
                                            knockout=FALSE,
                                            interval_col=&quot;timepoint&quot;,
                                            interval_step=2,
                                            tibble(expt = batch))
}, .x=batch))

wt_timepoint_pred_df = batches %&gt;% select(tp_preds) %&gt;% tidyr::unnest(tp_preds)

ggplot(wt_timepoint_pred_df, aes(x = timepoint)) +
  geom_line(aes(y = exp(log_abund) + exp(log_abund_detection_thresh), color=expt)) +
  facet_wrap(~cell_group, scales=&quot;free_y&quot;, nrow = 2) + monocle3:::monocle_theme_opts() + 
  ggtitle(&quot;wild-type kinetics by expt&quot;)

</code></pre>
<p><img alt="wt_expt_kinetics" src="../assets/wt_expt_kinetics.png" /></p>
<h5 id="perturbation-kinetics">Perturbation kinetics</h5>
<p>We can also plot a given perturbation's kinetics against the wild type. </p>
<pre><code>
tfap2afoxd3_cds = cds[, colData(cds)$gene_target %in% c(&quot;tfap2a-foxd3&quot;, control_ids)]

colData(tfap2afoxd3_cds)$gene_target = ifelse(colData(tfap2afoxd3_cds)$gene_target %in% control_ids, 
                                              &quot;control&quot;, colData(tfap2afoxd3_cds)$gene_target)

tfap2afoxd3_ccs = new_cell_count_set(tfap2afoxd3_cds, 
                                     sample_group = &quot;embryo&quot;, 
                                     cell_group = &quot;cell_type_sub&quot;)

colData(tfap2afoxd3_ccs)$timepoint = as.numeric(colData(tfap2afoxd3_ccs)$timepoint)


start_time = 18
stop_time = 72

time_formula = build_interval_formula(tfap2afoxd3_ccs, num_breaks = 3, interval_start = 18, interval_stop = 72)


tfap2afoxd3_ccm = new_cell_count_model(tfap2afoxd3_ccs, 
                                .     main_model_formula_str = paste0(time_formula, &quot;* perturbation&quot; ))

wt_timepoint_pred_df = estimate_abundances_over_interval(tfap2afoxd3_ccm, 
                                                                 interval_start=start_time, 
                                                                 interval_stop=stop_time, 
                                                                 interval_col=&quot;timepoint&quot;, 
                                                                 interval_step=2, 
                                                                 tibble(perturbation = &quot;control&quot;))

ko_timepoint_pred_df = estimate_abundances_over_interval(tfap2afoxd3_ccm, 
                                                                 interval_start=start_time, 
                                                                 interval_stop=stop_time, 
                                                                 interval_col=&quot;timepoint&quot;, 
                                                                 interval_step=2, 
                                                                 tibble(perturbation = &quot;tfap2a-foxd3&quot;))
timepoints = seq(start_time, stop_time, 2)
perturb_vs_wt_nodes = tibble(t1=timepoints) %&gt;%
  mutate(comp_abund = purrr::map(.f = compare_ko_to_wt_at_timepoint,
                                 .x = t1,
                                 perturbation_ccm=tfap2afoxd3_ccm,
                                 interval_col=&quot;timepoint&quot;,
                                 wt_pred_df = wt_timepoint_pred_df,
                                 ko_pred_df = ko_timepoint_pred_df)) %&gt;% 
  tidyr::unnest(comp_abund)

log_abund_detection_thresh=-3

ggplot(perturb_vs_wt_nodes, aes(x = t1)) +
  geom_line(aes(y = exp(log_abund_x) + exp(log_abund_detection_thresh), linetype = &quot;Wild-type&quot;)) +
  geom_line(aes(y = exp(log_abund_y) + exp(log_abund_detection_thresh), linetype = &quot;Knockout&quot;)) +
  ggh4x::stat_difference(aes(ymin = exp(log_abund_x)+exp(log_abund_detection_thresh), ymax = exp(log_abund_y) +exp(log_abund_detection_thresh)), alpha=0.3) + 
  facet_wrap(~cell_group, scales=&quot;free_y&quot;, nrow=2) + monocle3:::monocle_theme_opts() + ggtitle(&quot;tfap2afoxd3 kinetics&quot;)


</code></pre>
<p><img alt="perturb_kinetics" src="../assets/perturb_kinetics.png" /></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy 2023 <a href="https://github.com/cole-trapnell-lab"  target="_blank" rel="noopener">Cole Trapnell</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>