
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Maddy Duran">
      
      
        <link rel="canonical" href="https://cole-trapnell-lab.github.io/hooke/about/">
      
      
        <link rel="prev" href="../data/">
      
      
        <link rel="next" href="../advanced/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Differential Abundance - Hooke</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="trapnell" data-md-color-primary="indigo" data-md-color-accent="teal">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#differential-abundance" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Hooke" class="md-header__button md-logo" aria-label="Hooke" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Hooke
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Differential Abundance
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/cole-trapnell-lab/hooke" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Hooke" class="md-nav__button md-logo" aria-label="Hooke" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Hooke
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/cole-trapnell-lab/hooke" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../background/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Background
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_type/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data Type Requirements
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Example Datasets
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Differential Abundance
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Differential Abundance
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#analysis-of-differentially-abundant-cell-types-in-silicosis-with-hooke" class="md-nav__link">
    <span class="md-ellipsis">
      Analysis of differentially abundant cell types in silicosis with Hooke
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analysis of differentially abundant cell types in silicosis with Hooke">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#making-a-cell-count-set-object" class="md-nav__link">
    <span class="md-ellipsis">
      Making a cell count set object
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fitting-a-cell-count-model" class="md-nav__link">
    <span class="md-ellipsis">
      Fitting a cell count model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#estimating-abundances" class="md-nav__link">
    <span class="md-ellipsis">
      Estimating abundances
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#comparing-abundances" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing abundances
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting-differential-abundance-changes-on-a-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting differential abundance changes on a UMAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#controlling-for-batch-terms" class="md-nav__link">
    <span class="md-ellipsis">
      Controlling for batch terms
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../advanced/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Time Series Analysis
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#analysis-of-differentially-abundant-cell-types-in-silicosis-with-hooke" class="md-nav__link">
    <span class="md-ellipsis">
      Analysis of differentially abundant cell types in silicosis with Hooke
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analysis of differentially abundant cell types in silicosis with Hooke">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#making-a-cell-count-set-object" class="md-nav__link">
    <span class="md-ellipsis">
      Making a cell count set object
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fitting-a-cell-count-model" class="md-nav__link">
    <span class="md-ellipsis">
      Fitting a cell count model
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#estimating-abundances" class="md-nav__link">
    <span class="md-ellipsis">
      Estimating abundances
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#comparing-abundances" class="md-nav__link">
    <span class="md-ellipsis">
      Comparing abundances
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plotting-differential-abundance-changes-on-a-umap" class="md-nav__link">
    <span class="md-ellipsis">
      Plotting differential abundance changes on a UMAP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#controlling-for-batch-terms" class="md-nav__link">
    <span class="md-ellipsis">
      Controlling for batch terms
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="differential-abundance">Differential abundance</h1>
<p>Hooke can identify differentially abundant cell types for a given contrast. The underlying PLN model mimics the standard (G)LM-like interface of <code>R::stats</code>. This allows users to fit a multivariate Poisson lognormal model after correcting for effects of offsets and covariates. This framework allows users to perform multivariate statistical regression to describe how perturbations alter the relative abundances of each cell state. For example, you could model the effects of genotype (WT vs MT) while controlling for experimental batch. </p>
<h2 id="analysis-of-differentially-abundant-cell-types-in-silicosis-with-hooke">Analysis of differentially abundant cell types in silicosis with Hooke</h2>
<p>This data was published in <a href="https://www.biorxiv.org/content/10.1101/2023.02.17.528996v1">Hasegawa, Franks, et al. <em>bioRxiv</em></a>. 
The authors performed a longitudinal transcriptomic analysis using scRNA-sequencing of the lungs from pre- and post-silica challenged mice. </p>
<p>The data includes 12 whole lung samples across pre- and post- intratracheal silica. 35 unique cell states were identified using highly and specifically expressed marker genes. For simplicity, we are splitting samples into 2 levels: exposed and not exposed. The goal is to identify which cell types are differentially abundant post-silica exposure. </p>
<pre><code>cds = readRDS(&quot;silicosis_cds.rds&quot;)

# for simplicity, we are lumping together pre and post i.t. silica
colData(cds)$exposed = ifelse(colData(cds)$Timepoint == 0, &quot;not exposed&quot;, &quot;exposed&quot;)
colData(cds)$Rep = as.factor(colData(cds)$Rep)

plot_cells(cds, color_cells_by = &quot;fine_annotation&quot;, label_groups_by_cluster = F)

</code></pre>
<p><img alt="silicosis_umap" src="../assets/silcosis_umap.png" /></p>
<h3 id="making-a-cell-count-set-object">Making a cell count set object</h3>
<p>The first step of Hooke is to create a <code>cell_count_set</code> object. This requires three inputs: </p>
<ul>
<li><code>cds</code> - a monocle3 <code>cell_data_set</code> object</li>
<li><code>sample group</code> -  A column in <code>colData(cds)</code> that specifies how cells are grouped into samples</li>
<li><code>cell group</code> - A column in <code>colData(cds)</code> that specifies how cells are grouped into types or states (e.g. cluster)</li>
</ul>
<pre><code>ccs = new_cell_count_set(cds, 
                         sample_group = &quot;ID&quot;, 
                         cell_group = &quot;fine_annotation&quot;)
</code></pre>
<h3 id="fitting-a-cell-count-model">Fitting a cell count model</h3>
<p>The next step is to fit a <code>cell_count_model</code> given a model formula. This function works with the standard <code>R formula</code> notations. </p>
<p>A cell count model requires the following inputs: </p>
<ul>
<li><code>ccs</code> - a Hooke <code>cell_count_set</code> object</li>
<li><code>main_model_formula_str</code> -  A character string specifying the model of cell abundances across samples, where terms refer to columns in <code>colData(ccs)</code>. Put main effects here.</li>
<li><code>nuisance_model_formula_str</code> - A character string specifying the model of cell abundances across samples. Put nuisance effects here.</li>
</ul>
<p>In this case we are interested in comparing silica exposed mouse lungs to unexposed lungs. </p>
<pre><code>ccm  = new_cell_count_model(ccs,
                            main_model_formula_str = &quot;~ exposed&quot;)
</code></pre>
<h3 id="estimating-abundances">Estimating abundances</h3>
<p>Next we can predict fold change values for a given contrast of interest using <code>estimate_abundances()</code>.  </p>
<ul>
<li><code>ccm</code> - a Hooke <code>cell_count_model</code> object</li>
<li><code>newdata</code> - A tibble of variables used for the prediction.</li>
<li><code>min_log_abund</code> - numeric Minimum log abundance value.</li>
</ul>
<pre><code>cond_exp = estimate_abundances(ccm, tibble::tibble(exposed = &quot;exposed&quot;))
cond_not_exp = estimate_abundances(ccm, tibble::tibble(exposed = &quot;not exposed&quot;))

cond_not_exp %&gt;% head()
</code></pre>
<table>
<thead>
<tr>
<th>exposed</th>
<th>cell_group</th>
<th>log_abund</th>
<th>log_abund_se</th>
<th>log_abund_sd</th>
</tr>
</thead>
<tbody>
<tr>
<td>not exposed</td>
<td>Endothelial Arterial</td>
<td>2.87</td>
<td>0.196</td>
<td>0.559</td>
</tr>
<tr>
<td>not exposed</td>
<td>Endothelial Capillary</td>
<td>5.31</td>
<td>0.0591</td>
<td>0.559</td>
</tr>
<tr>
<td>not exposed</td>
<td>Endothelial Venous</td>
<td>3.09</td>
<td>0.179</td>
<td>0.559</td>
</tr>
<tr>
<td>not exposed</td>
<td>Lymphatic Endothelial</td>
<td>2.63</td>
<td>0.213</td>
<td>0.559</td>
</tr>
<tr>
<td>not exposed</td>
<td>Aerocytes</td>
<td>4.06</td>
<td>0.110</td>
<td>0.559</td>
</tr>
<tr>
<td>not exposed</td>
<td>Fibroblasts</td>
<td>5.26</td>
<td>0.0607</td>
<td>0.559</td>
</tr>
</tbody>
</table>
<h3 id="comparing-abundances">Comparing abundances</h3>
<p>And then compare two estimates of cell abundances from a Hooke model to identify significant changes with <code>compare_abundances()</code>. In this case, exposed vs not exposed. </p>
<ul>
<li><code>ccm</code> - a Hooke <code>cell_count_model</code> object</li>
<li><code>cond_x</code> - A cell type abundance estimate from estimate_abundances().</li>
<li><code>cond_y</code> - A cell type abundance estimate from estimate_abundances().</li>
</ul>
<pre><code>cond_ne_v_e_tbl = compare_abundances(ccm, cond_not_exp, cond_exp)

cond_ne_v_e_tbl %&gt;% select(cell_group, exposed_x, exposed_y,
                           delta_log_abund, delta_log_abund_se, delta_q_value)

</code></pre>
<table>
<thead>
<tr>
<th>cell_group</th>
<th>delta_log_abund</th>
<th>delta_log_abund_se</th>
<th>delta_q_value</th>
</tr>
</thead>
<tbody>
<tr>
<td>Endothelial Arterial</td>
<td>-0.584</td>
<td>0.196</td>
<td>0.0774</td>
</tr>
<tr>
<td>Endothelial Capillary</td>
<td>-0.400</td>
<td>0.0769</td>
<td>0.0131</td>
</tr>
<tr>
<td>Endothelial Venous</td>
<td>-0.353</td>
<td>0.145</td>
<td>0.130</td>
</tr>
<tr>
<td>Lymphatic Endothelial</td>
<td>0.0529</td>
<td>0.330</td>
<td>0.989</td>
</tr>
<tr>
<td>Aerocytes</td>
<td>-0.354</td>
<td>0.130</td>
<td>0.102</td>
</tr>
<tr>
<td>Fibroblasts</td>
<td>-0.477</td>
<td>0.102</td>
<td>0.0131</td>
</tr>
</tbody>
</table>
<h3 id="plotting-differential-abundance-changes-on-a-umap">Plotting differential abundance changes on a UMAP</h3>
<p>Finally, we can color our UMAP by the fold changes we estimate in <code>compare_abundances()</code> using the <code>plot_contrast()</code> function. We can select fold changes that are signifiant given a <code>q_value_thresh</code>. </p>
<ul>
<li><code>ccm</code> - A Hooke <code>cell_count_model</code> object.</li>
<li><code>cond_b_vs_a_tbl</code> - A data frame from <code>compare_abundances()</code>.</li>
<li><code>log_abundance_thresh</code> - <em>numeric</em> Select cell groups by log abundance.</li>
<li><code>q_value_thresh</code> - Remove contrasts whose change in q-value exceeds <code>q_value_thresh</code>.</li>
</ul>
<pre><code>plot_contrast(ccm, cond_ne_v_e_tbl, q_value_thresh = 0.05)
</code></pre>
<p><img alt="silicosis_rep" src="../assets/silicosis_no_rep.png" /></p>
<h3 id="controlling-for-batch-terms">Controlling for batch terms</h3>
<p>If your data contains multiple experimental batch, Hooke supports batch correction with the use of a nuisance model string term. </p>
<pre><code>ccm_rep  = new_cell_count_model(ccs,
                            main_model_formula_str = &quot;~ exposed&quot;,
                            nuisance_model_formula_str = &quot;~ Rep&quot;)

cond_exp_rep = estimate_abundances(ccm_rep, tibble::tibble(exposed = &quot;exposed&quot;, Rep = &quot;1&quot;))
cond_not_exp_rep = estimate_abundances(ccm_rep, tibble::tibble(exposed = &quot;not exposed&quot;, Rep = &quot;1&quot;))
cond_ne_v_e_tbl_rep = compare_abundances(ccm_rep, cond_not_exp_rep, cond_exp_rep)

plot_contrast(ccm_rep, cond_ne_v_e_tbl_rep, q_value_thresh = 0.05)
</code></pre>
<p>If we control for replicate, we see some difference in significant fold changes. </p>
<p><img alt="silicosis_rep" src="../assets/silicosis_rep.png" /></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      &copy 2023 <a href="https://github.com/cole-trapnell-lab"  target="_blank" rel="noopener">Cole Trapnell</a>

    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>